compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5189,6 +5191,262 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
    return resultReg;
    }
 
+static TR::Register *inlineSimpleAtomicUpdate(TR::Node *node, bool isAddOp, bool isLong, bool isGetThenUpdate, TR::CodeGenerator *cg)
+   {
+   TR::Node *valueAddrChild = node->getFirstChild();
+   TR::Node *deltaChild = NULL;
+
+   TR::Register *valueAddrReg = cg->evaluate(valueAddrChild);
+   TR::Register *deltaReg = NULL;
+   TR::Register *newValueReg = NULL;
+   TR::Register *oldValueReg = cg->allocateRegister();
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
+   TR::Register *resultReg = NULL;
+
+   bool isDeltaConstant = false;
+   bool isDeltaImmediate = false;
+   bool isDeltaImmediateShifted = false;
+   bool reuseDeltaReg = false;
+
+   int32_t numDeps = 3;
+
+   int32_t delta = 0;
+
+   // Memory barrier --- NOTE: we should be able to do a test upfront to save this barrier,
+   //                          but Hursley advised to be conservative due to lack of specification.
+   generateInstruction(cg, TR::InstOpCode::lwsync, node);
+
```

Comment:
I am wondering these symbols are only intended to map to the corresponding Unsafe methods.  If yes, the note on memory barriers is justified.  If not, the note is pessimistic regarding memory barriers.

Reply:
They are not. Unsafe is only one of many use cases of these symbols; other use cases include automatic reference count in Swift, as well as similar synchronization helpers for other languages.

Reply:
I spoke with Julian, and he indicated that his comment was about the use of memory barriers themselves rather than about the comment in the code.  He indicated that as these symbols are used for `Unsafe` (in addition to other uses) the memory barriers are required.